File: leetcode/0307-range-sum-query-mutable/solution.py
from typing import List


class NumArray:
    def __init__(self, nums: List[int]):
        self.n = len(nums)
        self.nums = nums[:]
        self.tree = [0] * (self.n + 1)
        for i, x in enumerate(nums):
            self._add(i, x)

    def _add(self, index: int, delta: int) -> None:
        i = index + 1
        while i <= self.n:
            self.tree[i] += delta
            i += i & (-i)

    def _prefix_sum(self, index: int) -> int:
        # sum of nums[0..index] inclusive
        total = 0
        i = index + 1
        while i > 0:
            total += self.tree[i]
            i -= i & (-i)
        return total

    def update(self, index: int, val: int) -> None:
        delta = val - self.nums[index]
        self.nums[index] = val
        self._add(index, delta)

    def sumRange(self, left: int, right: int) -> int:
        if left == 0:
            return self._prefix_sum(right)
        return self._prefix_sum(right) - self._prefix_sum(left - 1)


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# obj.update(index,val)
# param_2 = obj.sumRange(left,right)

File: leetcode/0307-range-sum-query-mutable/solution.cpp
#include <vector>
using namespace std;

class NumArray {
public:
    NumArray(vector<int>& nums) {
        n = nums.size();
        this->nums = nums;
        tree.assign(n + 1, 0);
        for (int i = 0; i < n; i++) {
            add(i, nums[i]);
        }
    }

    void update(int index, int val) {
        int delta = val - nums[index];
        nums[index] = val;
        add(index, delta);
    }

    int sumRange(int left, int right) {
        if (left == 0) {
            return prefixSum(right);
        }
        return prefixSum(right) - prefixSum(left - 1);
    }

private:
    int n;
    vector<int> nums;
    vector<int> tree;

    void add(int index, int delta) {
        int i = index + 1;
        while (i <= n) {
            tree[i] += delta;
            i += i & (-i);
        }
    }

    int prefixSum(int index) {
        int total = 0;
        int i = index + 1;
        while (i > 0) {
            total += tree[i];
            i -= i & (-i);
        }
        return total;
    }
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray* obj = new NumArray(nums);
 * obj->update(index,val);
 * int param_2 = obj->sumRange(left,right);
 */
